### PR TITLE
refactor[next]: workflowify PAST -> ITIR toolchain step

### PR DESCRIPTION
**Pull request by DropD**
## New:
- "next.ffront":
  - ".past_to_itir": workflow step PAST -> ITIR
  - ".transform_utils": some utility functions moved from "next.ffront.decorator"
 
## Changed:
 - "Backend":
   - is not a "ProgramProcessor" anymore
   - still passes "is_program_backend" test
   - new attribute ".transformer", a workflow which prepares for the ".executor"
     - currently starts at "stages.PastClosure" (PAST node + args, kwargs) but will evolve into starting from function definition + args, kwargs
 
 - "ModularExecutor":
   - is now a "ProgramExecutor" (passes "is_processor_kind" test)

 - "program_processor_interface":
   - "is_program_backend" test now checks for existence of an ".executor" attribute, which must pass the executor test
   
 - "ffront.decorator":
   - several utilities and methods refactored out into the new modules

- tests: test utils / fixtures refactored to deal with backends which are not executors

### CODE REVIEW

File: src/gt4py/next/ffront/decorator.py [hunk 1]
```diff
@@ -302,20 +245,41 @@ def __call__(self, *args, offset_provider: dict[str, Dimension], **kwargs) -> No
                 stacklevel=2,
             )
             with next_embedded.context.new_context(offset_provider=offset_provider) as ctx:
-                ctx.run(self.definition, *rewritten_args, **kwargs)
+                # TODO(ricoh): move into test
```
Comment (havogt) [inline src/gt4py/next/ffront/decorator.py:248]: I'd do this before this PR is merged. And then even split it into a separate PR.

**Review by havogt — CHANGES_REQUESTED** [review]
Looks good. A couple of things that I would move around and I'd split the past_to_func into a separate PR.

File: src/gt4py/next/otf/stages.py [hunk 2]
```diff
@@ -30,6 +32,15 @@
 SettingT_co = TypeVar("SettingT_co", bound=languages.LanguageSettings, covariant=True)
 
 
+@dataclasses.dataclass(frozen=True)
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/otf/stages.py:35]: Something like `ffront.transform_stages` might be an even closer fit.

File: src/gt4py/next/ffront/past_to_itir_wf.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# GT4Py - GridTools Framework
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/ffront/past_to_itir_wf.py:1]: Because that one is already huge and scrolling past the visitor is annoying. I could imagine something like

```
src/
  gt4py/
    next/
      ffront/
        past_to_itir/
          __init__.py
          visitor.py
          workflow.py  <-- in here?
```

File: src/gt4py/next/program_processors/runners/roundtrip.py [hunk 4]
```diff
@@ -227,8 +232,54 @@ def execute_roundtrip(
...
+    class Params:
+        roundtrip_workflow = factory.SubFactory(
+            RoundtripFactory, dispatch_backend=factory.SelfAttribute("..dispatch_backend")
+        )
+
+    dispatch_backend = None
+    otf_workflow = factory.LazyAttribute(lambda o: o.roundtrip_workflow)
+
+
+executor = RoundtripExecutorFactory(name="roundtrip")
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/program_processors/runners/roundtrip.py:280]: It makes it possible to create a double roundtrip with a customized roundtip inside without building everything explicitly:

```python
my_double_roundtrip = Backend(
    executor=roundtrip.RoundtripExecutorFactory(
        name="my_double_roundtrip",
        roundtrip_workflow__debug=True,
        roundtrip_workflow__lift_mode=my_lift_mode,
        dispatch_backend=RoundtripExecutorFactory(name="inner_roundtrip"),
    )
)
```

vs 

```python
my_dispatch_backend = RoundtripExecutor(
    name="inner_roundtrip",
    otf_workflow=roundtrip.Roundtrip(
        debug=True,
        lift_mode=my_lift_mode,
    ),
)

my_double_roundtrip = Backend(
    executor=roundtrip.RoundtripExecutor(
        name="my_double_roundtrip",
        otf_workflow=roundtrip.Roundtrip(
            debug=True,
            lift_mode=my_lift_mode,
            dispatch_backend=my_dispatch_backend,
        ),
        dispatch_backend=my_dispatch_backend
    )
)
```

File: src/gt4py/next/ffront/past_process_args_wf.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
+# GT4Py - GridTools Framework
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/ffront/past_process_args_wf.py:1]: In addition it contains the workflow step using the moved functions and packing their results into a `PastClosure`. Not worth spending much time on though.

File: src/gt4py/next/type_system/type_info.py [hunk 6]
```diff
@@ -754,3 +754,18 @@ def accepts_args(
         return True
 
     return next(errors, None) is None
+
+
+def flatten(arg: list | tuple | ts.TupleType):
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/type_system/type_info.py]: This will be in another PR, but I think the most elegant solution would be to make `TupleType` iterable and then simply use `eve.utils.flatten_iter`. This might actually be enough for all the other implementations too.

For reference: `primitive_constituents` doesn't work on sequences.

File: src/gt4py/next/ffront/past_to_func_wf.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+# GT4Py - GridTools Framework
```
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/ffront/past_to_func_wf.py]: Moved to #1487.

File: src/gt4py/next/type_system/type_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/type_system/type_info.py]: > was this moved? maybe add TODO with link to #1395

I commented on #1395 with the approach taken in #1487.

File: src/gt4py/next/otf/stages.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/otf/stages.py:35]: if you think it makes sense, go ahead, here or in a next PR.

File: src/gt4py/next/ffront/past_to_itir_wf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/ffront/past_to_itir_wf.py:1]: My view would be: make `ProgramLowering` private, then the only visible thing is the workflow. Personally, I wouldn't add an extra package.

Maybe, in a future refactoring, we could add a layer (maybe decorator) that makes an eve visitor a workflow. Currently we follow (in most places) the convention that a visitor/transformation has an apply which runs the transformation. The extra work in past_to_itir_wf could be part of the apply and the decorator would just turn `apply()` -> `__call__`. Just ideas.

Your choice how to leave the structure in this PR.

File: src/gt4py/next/program_processors/runners/roundtrip.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/program_processors/runners/roundtrip.py]: Thanks! Nice.

File: src/gt4py/next/type_system/type_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/type_system/type_info.py]: :+1:

File: src/gt4py/next/ffront/past_to_itir_wf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DropD) [reply src/gt4py/next/ffront/past_to_itir_wf.py:1]: I will pack it into the same module. I will not make the additional work in that step part of `ProgramLowering.apply()` in this PR because I can not judge if it would break a significant separation of concerns .
Also, for now I will refrain from renaming `ProgramLowering` for the following reasons:
- We have the convention that implementation details do not need to but are allowed to start with an underscore
- Importing or using `_ProgramLowering` in it's unit tests makes it seem like there is something tested that probably shouldn't be
- If after checking it turns out that the extra work required for the workflow step should indeed be part of `ProgramLowering`, then it would have to be undone again anyway.

**Review by havogt — APPROVED** [review]
Good first step! You need to merge with main to get CI green.

**Comment by havogt** [comment]
Maybe update the title before merge to be a bit more specific as it's only part of the workflowify
